File: contract/src/main/java/rrlane/leasing/contract/controller/LeasingContractController.java
package rrlane.leasing.contract.controller;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;
import rrlane.leasing.contract.dto.LeasingContractDTO;
import rrlane.leasing.contract.entity.LeasingContract;
import rrlane.leasing.contract.service.LeasingContractService;
import rrlane.leasing.util.Mapper;

import java.util.List;

@RequiredArgsConstructor
@RestController
@Slf4j
@RequestMapping("/contract")
public class LeasingContractController {
    private final LeasingContractService leasingContractService;

    // to create a vehicle leasing contract with POST and also to edit contract details with PATCH data
    @RequestMapping(method = {RequestMethod.POST, RequestMethod.PATCH})
    @CrossOrigin(origins = "http://localhost:4200/", maxAge = 3600)
    public ResponseEntity<LeasingContractDTO> updateContract(@RequestBody LeasingContractDTO contractDTO) {
        log.info("Sending contract details to server...");
        HttpStatus responseStatus = HttpStatus.OK;
        String responseFromService = leasingContractService.saveLeasingContract(contractDTO);
        if (responseFromService.contains("created")) responseStatus = HttpStatus.CREATED;
        log.info(responseFromService);
        return new ResponseEntity<>(contractDTO, responseStatus);
    }

    @GetMapping()
    @CrossOrigin(origins = "http://localhost:4200/", maxAge = 3600)
    public ResponseEntity<List<LeasingContractDTO>> viewContract() {
        log.info("Searching for policy contract on server...");
        List<LeasingContract> leasingContracts = leasingContractService.view();
        if (null == leasingContracts || 0 == leasingContracts.size()) {
            log.error("Policy contract could not be found on server.");
            return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
        }
        List<LeasingContractDTO> leasingContractDTOs = leasingContracts.stream().map(Mapper::entityToDto).toList();
        log.info("Found the contract details successfully in your name: " + leasingContractDTOs.get(0).getCustomerName());
        return new ResponseEntity<>(leasingContractDTOs, HttpStatus.OK);
    }
}

File: contractapi/src/main/kotlin/de/leasing/contract/api/controller/LeasingContractController.kt
package de.leasing.contract.api.controller

import de.leasing.contract.api.commons.Constants
import de.leasing.contract.api.entity.dto.LeasingContractReq
import de.leasing.contract.api.entity.dto.LeasingContractResp
import de.leasing.contract.api.service.LeasingContractService
import lombok.extern.slf4j.Slf4j
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.util.UriComponentsBuilder
import java.util.UUID

@RestController @RequestMapping("/api/v2/contracts/")
@Slf4j
class LeasingContractController(private val leasingContractService: LeasingContractService) {
    private val log: Logger = LoggerFactory.getLogger(LeasingContractController::class.java)

    @PostMapping
    fun createContract(
        @RequestBody leasingContractReq: LeasingContractReq,
        uriBuilder: UriComponentsBuilder
    ): ResponseEntity<LeasingContractResp> {
        log.info("Uploading contract details to server...")
        val newLeasingContract: LeasingContractResp = leasingContractService.saveLeasingContract(leasingContractReq)
        log.info(Constants.LEASING_CONTRACT_CREATED)

        val location = uriBuilder.path("/api/v2/contracts/{id}")
            .buildAndExpand(newLeasingContract.id).toUri()
        return ResponseEntity.created(location).body(newLeasingContract)
    }

    @GetMapping("{id}")
    fun viewContractById(
        @PathVariable() id: UUID
    ) : ResponseEntity<LeasingContractResp>{
        log.info("Fetching the contract record with the input ID...")
        return ResponseEntity.ok(leasingContractService.getLeasingContractById(id))
    }

    @GetMapping
    fun viewAllContracts(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
        @RequestParam(required = false) sortBy: String?
    ) : ResponseEntity<Iterable<LeasingContractResp>>{
        log.info("Fetching the list of all the contracts available on server...")
        return  if (null == sortBy) ResponseEntity.ok(leasingContractService.listAllLeasingContracts(page, size))
        else ResponseEntity.ok( leasingContractService.listAllLeasingContracts(sortBy))
    }
}